Frameworks/CoreGraphics/DWriteFontBinaryDataLoader.mm

```diff
@@ -0,0 +1,120 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import "DWriteFontBinaryDataLoader.h"
+
+using namespace Microsoft::WRL;
+
+/**
+ * Private class.
+ * Custom implementation of IDWriteFontFileStream that implements Read in terms of an underlying CFDataRef
+ *
+ * While IDWriteFontFileStream normally frees its underlying data bit by bit through ReleaseFileFragment(),
+ * for WinObjC purposes, it is easier to rely on existing mechanisms for CFData's destruction.
+ * Thus, this class releases its CFData all at once, at the time of its destruction.
+ */
+class DWriteFontBinaryDataStream : public RuntimeClass<RuntimeClassFlags<WinRtClassicComMix>, IDWriteFontFileStream> {
+protected:
+    InspectableClass(L"Windows.Bridge.DirectWrite", TrustLevel::BaseTrust);
+
+public:
+    DWriteFontBinaryDataStream() {
+    }
+
+    HRESULT RuntimeClassInitialize(CFDataRef data) {
+        CFRetain(data);
+        m_data.reset(data);
+
+        // Just use current time for m_lastWriteTime
+        FILETIME fileTime;
+        GetSystemTimeAsFileTime(&fileTime);
+
+        // Concat filetime into a single uint64_t
+        m_lastWriteTime = 0;
+        m_lastWriteTime |= static_cast<uint64_t>(fileTime.dwLowDateTime);
+        m_lastWriteTime |= static_cast<uint64_t>(fileTime.dwHighDateTime) << 32;
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE GetFileSize(_Out_ uint64_t* fileSize) {
+        RETURN_HR_IF_NULL(E_POINTER, fileSize);
+        *fileSize = CFDataGetLength(m_data.get());
+        return S_OK;
+    };
+
+    HRESULT STDMETHODCALLTYPE GetLastWriteTime(_Out_ uint64_t* lastWriteTime) {
+        RETURN_HR_IF_NULL(E_POINTER, lastWriteTime);
+        *lastWriteTime = m_lastWriteTime;
+        return S_OK;
+    };
+
+    // fragmentContext is deliberately unused: this is meant to be passed to ReleaseFileFragment() below to free part of the data,
+    // but this stream frees all the underlying CFData at once
+    HRESULT STDMETHODCALLTYPE ReadFileFragment(_Out_ const void** fragmentStart,
+                                               uint64_t fileOffset,
+                                               uint64_t fragmentSize,
+                                               _Out_ void** fragmentContext) {
+        if (fileOffset + fragmentSize > CFDataGetLength(m_data.get())) {
+            return E_INVALIDARG;
+        }
+
+        if (fragmentStart) {
+            const uint8_t* underlyingBuffer = CFDataGetBytePtr(m_data.get());
+            *fragmentStart = reinterpret_cast<const void*>(&(underlyingBuffer[fileOffset]));
+        }
+
+        if (fragmentContext) {
+            *fragmentContext = nullptr;
+        }
+
+        return S_OK;
+    };
+
+    void STDMETHODCALLTYPE ReleaseFileFragment(void* fragmentContext){
```

Comment:
it offers little by little reading which allows you to build releasable chunks no? This would improve memory consumption right?

---
In reply to: [90935815](https://github.com/Microsoft/WinObjC/pull/1504#discussion_r90935815) [](ancestors = 90935815,90916971,90747118,90743300)